sdk/python/lib/pulumi/x/automation/_stack.py

```diff
@@ -620,3 +670,45 @@ def fully_qualified_stack_name(org: str, project: str, stack: str) -> str:
     :returns: The fully qualified stack name.
     """
     return f"{org}/{project}/{stack}"
+
+
+def _create_log_file(command: str) -> Tuple[str, tempfile.TemporaryDirectory]:
+    log_dir = tempfile.TemporaryDirectory(prefix=f"automation-logs-{command}-")
+    filepath = os.path.join(log_dir.name, "eventlog.txt")
+
+    # Open and close the file to ensure it exists before we start polling for logs
+    f = open(filepath, "w+")
+    f.close()
+    return filepath, log_dir
+
+
+def _watch_logs(filename: str, callback: OnEvent):
+    with open(filename) as f:
+        while True:
+            line = f.readline()
+
+            # sleep if file hasn't been updated
+            if not line:
+                time.sleep(0.1)
+                continue
+
+            event = EngineEvent.from_json(json.loads(line))
+            callback(event)
+
+            # if this is the cancel event, stop watching logs.
+            if event.cancel_event:
+                break
+
+
+def _cleanup(temp_dir: Optional[tempfile.TemporaryDirectory],
+             thread: Optional[threading.Thread],
+             on_exit_fn: Optional[Callable[[], None]] = None) -> None:
+    # If there's an on_exit function, execute it (used in preview/up to shut down server)
```

Comment:
Sequencing seems to matter here if any of these throw exception. Let me try thread.join and see if that throws.

Reply:
Yeah seems so. If there's an exception in the background thread, it gets nicely printed out "Excpetion in thread Thread-1" with a stack trace. So far so good.

Reply:
Yes sounds like none of these finalizers will throw so we might be good here! The way to avoid having to reason/worry through this is nesting `with` statements and `from contextlib import contextmanager` - the with statements basically a condensed form of try..finally, that kind of code is easy to see at a glance that finalizers don't interfere and all try to run even if some fail.